rollup/fees/rollup_fee.go

```diff
@@ -138,3 +140,61 @@ func mulAndScale(x *big.Int, y *big.Int, precision *big.Int) *big.Int {
 	z := new(big.Int).Mul(x, y)
 	return new(big.Int).Quo(z, precision)
 }
+
+// copyTransaction copies the transaction, removing the signature
+func copyTransaction(tx *types.Transaction) *types.Transaction {
+	if tx.To() == nil {
+		return types.NewContractCreation(
+			tx.Nonce(),
+			tx.Value(),
+			tx.Gas(),
+			tx.GasPrice(),
+			tx.Data(),
+		)
+	}
+	return types.NewTransaction(
+		tx.Nonce(),
+		*tx.To(),
+		tx.Value(),
+		tx.Gas(),
+		tx.GasPrice(),
+		tx.Data(),
+	)
+}
+
+func CalculateTotalFee(tx *types.Transaction, state StateDB) (*big.Int, error) {
+	unsigned := copyTransaction(tx)
+	raw, err := rlpEncode(unsigned)
+	if err != nil {
+		return nil, err
+	}
+
+	l1BaseFee, overhead, scalar := readGPOStorageSlots(rcfg.L1GasPriceOracleAddress, state)
+	l1Fee := CalculateL1Fee(raw, overhead, l1BaseFee, scalar)
+
+	l2GasLimit := new(big.Int).SetUint64(tx.Gas())
+	l2Fee := new(big.Int).Mul(tx.GasPrice(), l2GasLimit)
+	fee := new(big.Int).Add(l1Fee, l2Fee)
+	return fee, nil
+}
+
+func VerifyFee(signer types.Signer, tx *types.Transaction, state StateDB) error {
+	fee, err := CalculateTotalFee(tx, state)
+	if err != nil {
+		return fmt.Errorf("invalid transaction: %w", err)
+	}
+
+	cost := tx.Value()
+	cost = cost.Add(cost, fee)
+	from, err := types.Sender(signer, tx)
+	if err != nil {
+		return errors.New("invalid transaction: invalid sender")
+	}
+	if state.GetBalance(from).Cmp(cost) < 0 {
+		return errors.New("invalid transaction: insufficient funds for gas * price + value")
+	}
```

Comment:
I see two risks with this:
1. An extra state query (`GetBalance`) per tx might have a performance impact.
2. On the client side, it's hard to debug the `insufficient funds` error when seemingly you can cover gas fee + value.

Otherwise it looks good, let's give it a try.